homeassistant/components/tuya/sensor.py

```diff
@@ -533,37 +535,46 @@ def __init__(
             elif self._status_range.type == "Enum":
                 self._type_data = EnumTypeData.from_json(self._status_range.values)
 
-        # Logic to ensure the set device class and API received Unit Of Measurement
-        # match Home Assistants requirements.
-        if (
-            self.device_class is not None
-            and not self.device_class.startswith(DOMAIN)
-            and description.native_unit_of_measurement is None
-        ):
-            # We cannot have a device class, if the UOM isn't set or the
-            # device class cannot be found in the validation mapping.
+        # For temperature sensors check whether TEMP_UNIT_CONVERT is available and, if so,
+        # use the unit of measurement specified by it
+        if description.key == DPCode.TEMP_CURRENT:
```

Comment:
I feel like you could rework this slightly, so that if the device doesn't have a TEMP_UNIT_CONVERT, and it is a temperature device, it defaults to Celcius?  (ie, C unless expressly marked as F)

Reply:
This is incorrect code, the integration will expose both sensors, there is no need to convert or add exceptions like these.

Reply:
1. The Tuya Climate platform has nearly identical [code ](https://github.com/home-assistant/core/blob/ba148863a78438716fd24371044d09811a8ee14e/homeassistant/components/tuya/climate.py#L152)

2. The API defines ```TEMP_UNIT_CONVERT``` for a reason. While the docs do mention that the unit for ```TEMP_CURRENT``` is Celsius this appears to be only the default unless TEMP_UNIT_CONVERT (which is also documented in many device instruction sets) is set to 'f'

3. The Kettle that I am actually using does not expose both DPCodes, it only exposes ```TEMP_CURRENT``` and ```TEMP_UNIT_CONVERT``` and ```TEMP_CURRENT``` is in Fahrenheit. See the device status Json I posted in PR description

4. I am fine with reporting a bug upstream but I am not sure that's going to go anywhere. While we wait for that to be (not) fixed, what exactly is the issue with having a client-side workaround so that folks can at least get their devices working properly?

Reply:
> The Tuya Climate platform has nearly identical code

That platform can't handle multiple sensors, so it has to figure it out for that reason. This is not applicable to the sensor platform.

> The Kettle that I am actually using does not expose both DPCodes, it only exposes TEMP_CURRENT and TEMP_UNIT_CONVERT and TEMP_CURRENT is in Fahrenheit

`TEMP_CURRENT` is already exposed.

![image](https://user-images.githubusercontent.com/195327/139723975-26d3335f-4774-49e0-8e41-34f71e328b48.png)

I'm not sure what is wrong at this point? The data point exposes its unit in the API, which is used.

![image](https://user-images.githubusercontent.com/195327/139724103-13fd914c-029a-47b6-9f9c-475d655e4a3d.png)

So, what is the issue you are fixing right now, because the is something I'm now missing?

Reply:
> I'm not sure what is wrong at this point?
```TEMP_CURRENT``` is exposed but it's unit of measurement is set to Celsius. Device is actually returning temperature in Fahrenheit and ```TEMP_UNIT_CONVERT``` is set to F.

So, yes, the data point unit exposed from the API is incorrect but it's likely a device\firmware issue not an API issue.

I am still not clear about why this was considered 'ok' for the climate platform which had to deal with exactly the same issue for the current temperature. This has nothing to do with multiple sensors, it has to do with exposing the right unit of measurement for a single sensor and doing so requires querying another data point from the device.

> Well, that approach for not trying because we think it won't be handle
It's not mutually exclusive. I am not exactly sure what's wrong with having a workaround in the code and removing if\when upstream is fixed.

Reply:
Ok, so the unit returned by the API is just plain dead wrong. This needs to be resolved upstream, not in the integration.

This is absolute not the same as in the climate entity. In the climate entity the logic is used to decide which DP to use, in this PR it is worked around to override an incorrect DP.

Reply:
Ummm, here is the code from the climate entity, it's most definitely choosing the unit of measurement and not deciding which DP to use
![Screenshot 2021-11-01 170422](https://user-images.githubusercontent.com/44785744/139758372-6de85fa6-6faa-4d4c-8b58-ded6a1e64c90.png)

Anyway, whatever, I give up :)

Cheers!

Reply:
That same logic in another place as well, choosing the control that goes with it.
For climate, this is a different case.

>  I give up :)

Please don't. Reach out to tuya, report the DP returns a mismatching unit of measurement.